In [1]:
### Run this cell before continuing. 
library(tidyverse)
library(repr)
library(tidymodels)
library(janitor)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
── Attaching packages ────────────────────────────────────── tidymodels 1.1.1 ──

✔ broom        1.0.6     ✔ rsample      1.2.1
✔ dials        1.3.0     ✔ tune         1.1.2
✔ infer        1.0.7     ✔ workflows    1.1.4
✔ modeldata    1.4.0     ✔ workflowsets 1.0.1
✔ parsnip      1.2.1     ✔ yardstick    1.3.1
✔ recipes      1.1.0     

── Conflicts ───────────────────────────────────────── tidymodels_conflicts() ──
✖ scales::discard() masks purrr::discard()
✖ dplyr::filt

In [2]:
players <- read_csv("https://github.com/maru-thedog/dsci-100_group_project/raw/refs/heads/main/players.csv")
mean_played_hours <- mean(players$played_hours, na.rm = TRUE)
mean_age <- mean(players$Age, na.rm = TRUE)
mean_played_hours
mean_age

sessions <- read_csv("https://github.com/maru-thedog/dsci-100_group_project/raw/refs/heads/main/sessions.csv")
sessions <- sessions |> 
            clean_names() |>
            mutate(start_time = dmy_hm(start_time)) |>
            mutate(end_time = dmy_hm(end_time)) |>
            select(-starts_with("original")) 
head(players)

Rows: 196 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): experience, hashedEmail, name, gender
dbl (2): played_hours, Age
lgl (1): subscribe

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


[1] 5.845918

[1] 20.52062

Rows: 1535 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): hashedEmail, start_time, end_time
dbl (2): original_start_time, original_end_time

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


experience subscribe
1 Pro         TRUE    
2 Veteran     TRUE    
3 Veteran    FALSE    
4 Amateur     TRUE    
5 Regular     TRUE    
6 Amateur     TRUE    
  hashedEmail                                                      played_hours
1 f6daba428a5e19a3d47574858c13550499be23603422e6a0ee9728f8b53e192d 30.3        
2 f3c813577c458ba0dfef80996f8f32c93b6e8af1fa939732842f2312358a88e9  3.8        
3 b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3c5a9d2118eb7ccbb28  0.0        
4 23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4fa7a5a659ff443a0eb5  0.7        
5 7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb0af4d48fcce2420f3e  0.1        
6 f58aad5996a435f16b0284a3b267f973f9af99e7a89bee0430055a44fa92f977  0.0        
  name      gender Age
1 Morgan    Male    9 
2 Christian Male   17 
3 Blake     Male   17 
4 Flora     Female 21 
5 Kylie     Male   21 
6 Adrian    Female 17

# Data Description
## Players Data : There are 196 observations and 7 variables
* experience : player's experience level
* subscribe : they subscribe game newsletter or not
* hashedEmail : hash of player's email  
* played_hurs : total hours of game play
* name : player's name
* gender : player's gender
* Age : player's age

## Sessons Data : There are 1535 observations and 5 variables
* hashedEmail : hash of player's email   
* start_time : start time of the game session
* end_time : end time of the game session
* original_start_time : Unix timestamp for start_time (by searching)
* original_end_time : Unix timestamp for end_time (by searching)

## Issues
* some age values are missing, which is very important in analyzing a game related data


# Questions
* Can player's experience and age predict played hours?

# Exploratory Data Analysis and Visualization
| Played Hours (mean) | Age(mean) |
| -------- | ------- |
| 5.85 | 20.52 |

In [3]:
# played_hours_histogram <- ggplot(players, aes(x = played_hours)) + 
#                             geom_histogram(binwidth = 5, fill = "green", color = "black", alpha = 0.5) +
#                             labs(title = "Distribution of Played Hours", x = "Played Hours", y = "Frequency")
# played_hours_histogram

# experience_played_hours_plot <- ggplot(players, aes(x = experience, y = played_hours, fill = experience)) +
#                                 geom_boxplot() +
#                                 labs(title = "Played Hours by Experience Level", x = "Experience Level", y = "Played Hours")
# experience_played_hours_plot

# age_played_hours_plot <- ggplot(players, aes(x = Age, y = played_hours)) +
#                         geom_point(color = "blue", alpha = 0.5) +
#                         labs(title = "Scatter Plot of Age vs. Played Hours", x = "Age", y = "Played Hours") 
# age_played_hours_plot

# Methods and plan
* I choose linear regression method. we can use expericen or age as predictor and played hours as response variable
* Assume that there is a linear relationship between predictor and response variable
* limitations : if the relationship is not linear, we will fail
* split data : 75% training and 25% testing / and cross-validation
* need to normalilze the values
* remove missing data in Age
* have to convert experience into numerial valeus

# Group part Group part Group part Group part Group part Group part Group part Group part 
# Group part Group part Group part Group part Group part Group part Group part Group part 
# Group part Group part Group part Group part Group part Group part Group part Group part 
# Group part Group part Group part Group part Group part Group part Group part Group part 
# Group part Group part Group part Group part Group part Group part Group part Group part 

# Broad question : Question 2: We would like to know which "kinds" of players are most likely to contribute a large amount of data so that we can target those players in our recruiting efforts.
# Specific question : Can player's experience and age predict played hours? (knn)

In [20]:
players <- players |>
                mutate(numerical_experience = case_when (experience == "Beginner" ~ 1, experience == "Regular" ~ 2, experience == "Amateur" ~ 3, experience == "Pro" ~ 4, experience == "Veteran" ~ 5),
                      played_hours = as.factor(played_hours), experience = as.factor(experience))


# #use experience to predict played hours
# experience_split <- initial_split(players, prop = 0.75, strata = played_hours)  
# experience_train <- training(experience_split)   
# experience_test <- testing(experience_split)

# #experice -> played hours recipe
# experience_recipe <- recipe(played_hours ~ numerical_experience , data = experience_train) |>
#                    step_scale(all_predictors()) |>
#                    step_center(all_predictors())

# #experice -> played hours knn specification from k=4, accuracies are same 
# experience_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = 4) |>
#       set_engine("kknn") |>
#       set_mode("classification")

############################################################################################################
#checking best k value -> 4 is the best k value
# experience_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) |>
#       set_engine("kknn") |>
#       set_mode("classification")

# experience_folds <- vfold_cv(experience_train, v = 5, strata= played_hours)

# knn_grid <- tibble(neighbors = c(2, 3, 4, 5, 6))

# knn_results <- workflow() |>
#                 add_recipe(experience_recipe) |>
#                 add_model(experience_spec) |>
#                 tune_grid(resamples = experience_folds, grid = knn_grid) |>
#                 collect_metrics()
# accuracies <- knn_results |>
#             filter(.metric == "accuracy") 
# cross_val_plot <- ggplot(accuracies, aes(x = neighbors, y = mean)) +
#                     geom_point() +
#                     geom_line() +
#                     labs(x = "Number of Neighbors", y = "Accuracy")
# cross_val_plot
############################################################################################################

# experience_fit <- workflow() |>
#       add_recipe(experience_recipe) |>
#       add_model(experience_spec) |>
#       fit(data = experience_train)

# experience_test_predictions <- predict(experience_fit , experience_test) |>
#       bind_cols(experience_test)

# experience_prediction_accuracy <- experience_test_predictions |>
#                                     metrics(truth = played_hours, estimate = .pred_class)
# experience_prediction_accuracy

################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################

# use age to predict played hours
age_split <- initial_split(players, prop = 0.75, strata = Age)  
age_train <- training(age_split)   
age_test <- testing(age_split)

#experice -> played hours recipe
age_recipe <- recipe(played_hours ~ Age, data = age_train) |>
   step_scale(all_predictors()) |>
   step_center(all_predictors())

age_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = 9) |>
      set_engine("kknn") |>
      set_mode("classification")

############################################################################################################
#checking best k value -> 9 is the best k value
# age_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) |>
#       set_engine("kknn") |>
#       set_mode("classification")

# age_folds <- vfold_cv(age_train, v = 9, strata= played_hours)

# knn_grid <- tibble(neighbors = c(2, 3, 4, 5, 6, 7, 8, 9, 10))

# knn_results <- workflow() |>
#                 add_recipe(age_recipe) |>
#                 add_model(age_spec) |>
#                 tune_grid(resamples = age_folds, grid = knn_grid) |>
#                 collect_metrics()
# accuracies <- knn_results |>
#             filter(.metric == "accuracy") 
# cross_val_plot <- ggplot(accuracies, aes(x = neighbors, y = mean)) +
#                     geom_point() +
#                     geom_line() +
#                     labs(x = "Number of Neighbors", y = "Accuracy")
# cross_val_plot
############################################################################################################

#experice -> played hours knn specification
age_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = 3) |>
              set_engine("kknn") |>
              set_mode("classification")

age_fit <- workflow() |>
      add_recipe(age_recipe) |>
      add_model(age_spec) |>
      fit(data = age_train)

age_test_predictions <- predict(age_fit , age_test) |>
      bind_cols(age_test)

age_prediction_accuracy <- age_test_predictions |>
                            metrics(truth = played_hours, estimate = .pred_class)
age_prediction_accuracy


################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################
################################################################################################################


.metric  .estimator .estimate  
1 accuracy multiclass 0.300000000
2 kap      multiclass 0.009620826